src/dataflow/src/sink/tail.rs

```diff
@@ -94,14 +101,74 @@ pub fn tail<G>(
                 return;
             }
         });
+
+        let progress_row = update_progress(
+            &mut input_frontier,
+            input.frontier().frontier(),
+            &mut packer,
+            connector.object_columns + 1,
+        );
+
+        // Only emit updates if this operator/worker received actual
+        // data for emission. For TAIL, data is exchanged to one worker,
+        // which forwards all the data to the client process. If we
+        // blindly forwarded the frontier from all workers we would get
+        // multiple progress updates in the client.
+        if connector.emit_progress && received_data {
+            if let Some(progress_row) = progress_row {
+                let results = vec![progress_row];
+
+                // TODO(benesch): the lack of backpressure here can result in
+                // unbounded memory usage.
+                if connector.tx.send(results).is_err() {
+                    errored = true;
+                    return;
+                }
+            }
+        }
     })
 }
 
-fn batch_upper(
-    batch: Option<&Rc<OrdValBatch<GlobalId, Row, u64, isize, usize>>>,
-) -> Option<Timestamp> {
-    batch
-        .map(|b| b.desc.upper().elements().get(0))
-        .flatten()
-        .copied()
+// Checks if there is progress between `current_input_frontier` and
+// `new_input_frontier`. If yes, updates the `current_input_frontier` to
+// `new_input_frontier` and returns a [`Row`] that encodes this progress. If
+// there is no progress, returns [`None`].
+fn update_progress(
+    current_input_frontier: &mut Antichain<Timestamp>,
+    new_input_frontier: AntichainRef<Timestamp>,
+    packer: &mut Row,
+    empty_columns: usize,
+) -> Option<Row> {
+    // Test to see if strict progress has occurred. This is true if the new
+    // frontier is not less or equal to the old frontier.
+    let progress = !PartialOrder::less_equal(&new_input_frontier, &current_input_frontier.borrow());
+
+    if progress {
+        current_input_frontier.clear();
+        current_input_frontier.extend(new_input_frontier.iter().cloned());
```

Comment:
Can you do `current_input_frontier = new_input_frontier.to_owned()` and remove the .clear? The use of `to_owned` can replace the `.iter().cloned()` at the least.

Reply:
`to_owned` will not reuse the allocation which I think is the goal of the current implementation

Reply:
This pattern is used in parts of differential/timely that's why I'm using it here as well. `new_input_frontier.to_owned()` would do a `to_vec()` internally which allocates a new `Vec` and clone the values. Just doing the `iter().cloned()` will only clone the values and reuse the existing `Vec` in `current_input_frontier`.

Plus, here `current_input_frontier` is a `&mut` parameter so I would have to do some re-shuffling of the code.

I'm happy to do it if you think it's worthwhile after the above explanations.

Reply:
TIL! Thanks.

Reply:
You might be able to do `current_input_frontier.clone_from(&new_input_frontier)`, which will re-use the allocation, and any allocations involved with the timestamp as well.

Reply:
You could probably also defer the lines to the bottom of the scope and just assign there:
```
*current_input_frontier = new_input_frontier;
```
unless I totally misunderstand the logic.

Reply:
I don't see how those would work. `current_input_frontier` is a `&mut Antichain` but I do hold the `Antichain` outside the method call. And `new_input_frontier` is an `AntichainRef`, which is the common thing I can get from the two call sites.

I could easily be missing something, of course. :sweat_smile:

Reply:
> And new_input_frontier is an AntichainRef,

Ah, this makes sense sorry. I saw that it came in as owned, and my eyes glazed over the "no lifetime" part of the type and I thought it would just be easy.

Reply:
No worries! :sweat_smile: